src/osdc/ObjectCacher.cc

```diff
@@ -2526,11 +2526,6 @@ void ObjectCacher::_discard_finish(ObjectSet *oset, bool was_dirty,
 {
   assert(lock.is_locked());
 
-  // did we truncate off dirty data?
-  if (flush_set_callback && was_dirty && oset->dirty_or_tx == 0) {
-    flush_set_callback(flush_set_callback_arg, oset);
-  }
-
```

Comment:
I think the correct fix should be 
<pre>
diff --git a/src/osdc/ObjectCacher.cc b/src/osdc/ObjectCacher.cc
index 50f9a8d6a6..1049863815 100644
--- a/src/osdc/ObjectCacher.cc
+++ b/src/osdc/ObjectCacher.cc
@@ -2489,8 +2489,8 @@ void ObjectCacher::discard_writeback(ObjectSet *oset,
 
   if (gather.has_subs()) {
     gather.set_finisher(new FunctionContext(
-      [this, oset, was_dirty, on_finish](int) {
-        _discard_finish(oset, was_dirty, on_finish);
+      [this, oset, on_finish](int) {
+        _discard_finish(oset, false, on_finish);
       }));
     gather.activate();
     return;
</pre>

Reply:
Actually, we don't need the second parameter 'was_dirty' and it's not used in function _discard_finish. I think we should delete this parameter and change the place where call the 
_discard_finish function crrespondingly, is that okay?

Reply:
no. if gather.has_subs() is false, _discard_finish() needs to call flush_set_callback

Reply:
I think the FILE_CACHE | FILE_BUFFER caps only should be dropped by the write callback. If we drop it manually will it cause a double drop? This has happened in my experimental environment.

Reply:
I think the FILE_CACHE | FILE_BUFFER caps only should be dropped by the write callback. If we drop it manually will it cause a double drop? This has happened in my experimental environment. A double drop of FILE_CACHE | FILE_BUFFER will also lead a assert fail.

Reply:
if ObjectCacher::discard_writeback() drop all dirty buffers and there is no writeback buffer, who calls flush_set_callback()

Reply:
ok, you are right, i ignored this case.

Reply:
Do I need to combine these two commits into one?

Reply:
yes